### ISSUE TITLE
Nesting no_grad in autocast causes backwards graph to be (partially) lost outside of no_grad

### ISSUE THREAD

**Issue by haydn-jones** [pytorch/pytorch#112583]
Hmm, I actually think your example is unexpected, can you file a separate bug for it.

_Originally posted by @ezyang in https://github.com/pytorch/pytorch/issues/105211#issuecomment-1787772959_

If I'm in an autocast context and I do a forward pass of a model in no_grad, then do another forward pass outside of that no_grad context but still in the autocast context, arbitrary nodes in the graph will be lost. See example:

```Python
import torch
import torch.nn as nn

l1 = nn.Sequential(
    nn.Linear(2, 2),
).cuda()

l2 = nn.Sequential(
    nn.Linear(2, 2),
    nn.LayerNorm(2),
).cuda()

x = torch.randn(2, 2).cuda()

#################################
# Just linear
################################
with torch.cuda.amp.autocast():
    with torch.no_grad():
        y1 = l1(x)

    y1_2 = l1(x)
    print(y1_2.grad_fn) # None

# Remove autocast
with torch.no_grad():
    y1 = l1(x)

y1_2 = l1(x)
print(y1_2.grad_fn) # AddmmBackward

#################################
# Linear -> LayerNorm makes output
# have grad_fn
################################
with torch.cuda.amp.autocast():
    with torch.no_grad():
        y2 = l2(x)

    y2_2 = l2(x)
    print(y2_2.grad_fn) # LayerNormBackward

with torch.no_grad():
    y2 = l2(x)

y2_2 = l2(x)
print(y2_2.grad_fn) # Still LayerNormBackward

with torch.cuda.amp.autocast():
    with torch.no_grad():
        y2 = l2(x)

    y2_2 = l2(x)
    print(y2_2.grad_fn) # LayerNormBackward

y2_2.sum().backward()
print(l2[0].weight.grad) # None
print(l2[1].weight.grad) # Not none
```

PyTorch version: 2.1.0

**Comment by nicolas-dufour** [comment]
I have the same issue with Pytorch 2.0. 
Some related issues: https://discuss.pytorch.org/t/autocast-and-torch-no-grad-unexpected-behaviour/93475
https://github.com/pytorch/pytorch/issues/65766

**Comment by haydn-jones** [comment]
Interesting, thanks for the link. It looks like there is a question of whether or not it is a [bug](https://discuss.pytorch.org/t/autocast-and-torch-no-grad-unexpected-behaviour/93475/3), though if it's not considered a bug the behavior should at least be consistent IMO. Network `l1` in my example will produce an error if you try to run backward on some loss, while `l2` will not produce an error simply because a LayerNorm was introduced (though only the layer norms will receive gradients in `l2`).

**Comment by ezyang** [comment]
oh you're right, this is a dupe of https://github.com/pytorch/pytorch/issues/65766

**Comment by ptrblck** [comment]
You can also disable the cache as a workaround if it's blocking you and the suggestions in the linked post don't work.

**Comment by albanD** [comment]
A simpler option than having a  grad_mode depdendent cache is to invalidate the cache when the grad_mode is detected to be different from what it was before.

We would accept a PR fixing this!

**Comment by haydn-jones** [comment]
Alright, will give fixing it a shot. May come back with some questions.

**Comment by albanD** [comment]
This is being fixed by https://github.com/pytorch/pytorch/issues/112583

**Comment by albanD** [comment]
FYI https://github.com/pytorch/pytorch/pull/113685 got stale but feel free to pick it back up!

**Comment by keyboardAnt** [comment]
With torch version 2.4.0:
```python
import torch

x = torch.tensor([1.], requires_grad=True)
print(f"1. {x.requires_grad=}")

with torch.no_grad():
    y = x * 2
    print(f"2. {y.requires_grad=}")

print(f"3. {y.requires_grad=}")

@torch.no_grad()
def f():
    x = torch.tensor([1.], requires_grad=True)
    print(f"4. {x.requires_grad=}")
    y = x * 2
    print(f"5. {y.requires_grad=}")
    g(x, y)

def g(x, y):
    print(f"6. {x.requires_grad=}")
    print(f"7. {y.requires_grad=}")
    w = x * y
    print(f"8. {w.requires_grad=}")
    z = x * x
    print(f"9. {z.requires_grad=}")


f()
```
prints:
```text
1. x.requires_grad=True
2. y.requires_grad=False
3. y.requires_grad=False
4. x.requires_grad=True
5. y.requires_grad=False
6. x.requires_grad=True
7. y.requires_grad=False
8. w.requires_grad=False
9. z.requires_grad=False
```

**Comment by barakugav** [comment]
+1
